zephyr hfp hf sal

expliyh opened the pull request:
depends-on: [ open-vela/vendor_openvela/pull/30 ]

gzh-terry commented on feature/src/system_bluetooth_bt_impl.c:
```diff
@@ -66,7 +66,7 @@ static bt_status_t bluetooth_connect_profiles(FeatureInstanceHandle feature, bt_
     case HFP_HF:
 #ifdef CONFIG_BLUETOOTH_HFP_HF
-        status = bt_hfp_hf_connect(bt_ins, addr);
+        status = bt_client_hfp_hf_connect(bt_ins, addr);
```
Do not change the names for framework APIs.

gzh-terry commented on service/stacks/include/sal_hfp_hf_interface.h:
```diff
@@ -0,0 +1,48 @@
+#include <stdint.h>
+
+#include "bt_addr.h"
+#include "bt_hfp_hf.h"
```
Remove `#include "bt_hfp_hf.h"`

expliyh replied:
Done

gzh-terry commented on service/stacks/include/sal_hfp_hf_interface.h:
```diff
@@ -0,0 +1,48 @@
+#include "bt_addr.h"
+#include "bt_hfp_hf.h"
+#include "bt_status.h"
+#include "hfp_hf_service.h"
```
Remove `#include "hfp_hf_service.h"`

gzh-terry replied:
Add `bt_hfp.h` instead.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+uint8_t on_sdp_done(struct bt_conn *conn, struct bt_sdp_client_result *result, const struct bt_sdp_discover_params *ignore);
+
+NET_BUF_POOL_DEFINE(sdp_discover_pool, 10, BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
```
`10` -> `CONFIG_BT_MAX_CONN`

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+static struct bt_sdp_discover_params sdp_discover = {
+	.func = on_sdp_done,
+	.pool = &sdp_discover_pool,
```
Let's follow the shell from zephyr?
```
static struct bt_sdp_discover_params sdp_discover = {
	.type = BT_SDP_DISCOVER_SERVICE_SEARCH_ATTR,
	.uuid = BT_UUID_DECLARE_16(BT_SDP_HANDSFREE_SVCLASS),
	.func = on_sdp_done,
	.pool = &sdp_discover_pool,
};
```

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+typedef struct _bt_hfp_hf_connection {
+    bt_address_t* addr;
+    struct bt_conn* conn;
+    struct bt_list_t* incoming_calls;
```
`calls`, not only incoming calls.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+};
+
+typedef struct _bt_hfp_hf_connection {
+    bt_address_t* addr;
```
It is not necessary to store an address if it belongs to `*conn`.

Otherwise, we would always need to clear `*addr` when `*conn` is released. This would introduce a risk if we forget to do so.

Furthermore, if we really need an address for a quick operation, we should store the `addr` itself rather than a pointer, as it only uses 6 bytes.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    
+    if(bt_hfp_hf_connect(conn, &hf, channel)){
+        BT_LOGE("%s, Failed to initiate HFP HF connection", __func__);
+        bt_conn_unref(conn);
```
No need to `bt_conn_unref(conn)` here.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    struct bt_conn* conn;
+    struct bt_list_t* incoming_calls;
+    struct bt_hfp_hf *hf;
+} bt_hfp_hf_connection_t;
```
BTW, we might need a `struct bt_conn *sco_conn;` in the future.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+}
+
+static struct bt_hfp_hf_cb hf_callbacks = {
+    .connected = on_hfp_hf_connected,
```
Keep function names consistent,
```
.connected = zblue_on_hf_connected,
.disconnected = zblue_on_hf_disconnected,
.xxxxx = zblue_on_hf_xxxxx,
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    .subscriber_number = NULL,
+};
+
+bt_status_t bt_sal_hfp_hf_init(uint32_t hf_features, uint8_t p_max_connection)
```
Remove `p_`, this is not a pointer.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    printf("bt_sal_hfp_hf_cleanup: Currently not supported\n");
+}
+
+bt_status_t pre_hfp_hf_connect()
```
Remove this later

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+    return BT_STATUS_SUCCESS;
+}
+
```
Remove extra line.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+    if (!conn){
+        BT_LOGW("%s, acl not conneted, try connect\n", __func__);
+        if (bt_sal_connect(0, addr) != BT_STATUS_SUCCESS)
```
Remove these `pre_connect`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+            BT_LOGE("%s, Failed to allocate memory for new HFP HF connection", __func__);
+            return BT_STATUS_FAIL;
+        }
+        bt_sal_get_remote_address(conn, new_connection->addr);
```
Caution, `new_connection->addr` is a `nullptr` at this moment.

gzh-terry replied:
BTW, when `bt_sal_get_remote_address` fails, do not store `conn`, and disconnect `hf` immediately.

expliyh replied:
Code changed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    int err;
+    uint16_t value;
+
+    BT_LOGD("Discover done");
```
Say something like "HFP-HF discovered"

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+        conn = bt_conn_lookup_addr_br(addr);
+    }
+
+    SAL_CHECK_RET(bt_sdp_discover(conn, &sdp_discover), 0);
```
What if the remote device does not support AG?

expliyh replied:
Now we will disconnect profile when sdp fails

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+static void free_connection(void* p_data)
+{
+    bt_hfp_hf_connection_t* data = (bt_hfp_hf_connection_t*)p_data;
+    bt_conn_unref(data->conn);
```
Remove `bt_conn_unref`, since we do not `bt_conn_ref` when adding a node to the list.

expliyh replied:
We got the ref when connected callback is called, and unref when disconnect(sal conn remove)

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+bt_status_t bt_sal_hfp_hf_init(uint32_t hf_features, uint8_t p_max_connection)
+{
+    max_connection = p_max_connection;
```
Didn't see this `max_connection`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+
+bt_status_t bt_sal_hfp_hf_connect(bt_address_t* addr)
+{
+    struct bt_conn* conn = bt_conn_lookup_addr_br(addr);
```
For each `bt_conn_lookup_addr_br` remember to `bt_conn_unref(conn);`

For example, before `return BT_STATUS_FAIL;`, and when `bt_sdp_discover` fails, and when `do_hf_connect` completed.

BTW, I would recommend not to check SDP at this moment. We can first assume that the remote device support HFP-AG.

We can check SDP unpon ACL connection in the future, which is not related to HFP it self.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+        BT_LOGW("%s, acl not conneted, try connect\n", __func__);
+        if (bt_sal_connect(0, addr) != BT_STATUS_SUCCESS)
+            return BT_STATUS_FAIL;
+        conn = bt_conn_lookup_addr_br(addr);
```
Same here, always remember to `bt_conn_unref(conn)` for each `bt_conn_lookup_addr_br()`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+    }
+}
+
+bt_status_t do_hf_connect(struct bt_conn *conn, uint16_t channel) {
```
`static`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+NET_BUF_POOL_DEFINE(sdp_discover_pool, 10, BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
+		    CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
+
+static struct bt_sdp_discover_params sdp_discover = {
```
`static const`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,358 @@
+ * limitations under the License.
+ ***************************************************************************/
+
+#include "sal_hfp_hf_interface.h"
```
In alphabetic order, i.e.,
```
#include <stdio.h>

#include "bt_debug.h"
#include "sal_connection_manager.h"
#include "sal_hfp_hf_interface.h"
#include "sal_interface.h"
#include "sal_zblue.h"
```

expliyh committed:
zblue: add sal API declarations

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: add sal hfp C file

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: add empty HFP HF implementation

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: add empty HFP AG  implementation

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: add connections list to store connection info

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement hf connect

bug: v/77997

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement incoming call callback

bug: v/77999

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement ATA

bug: v/78000

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement ATD API

bug: v/78000

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/include/sal_hfp_internal.h:
```diff
@@ -0,0 +1,25 @@
+
+#include "service_loop.h"
+
+#define Z_API(func) z_##func
```
Use one `#define` for both SAL and zephyr.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+static void free_connection(void* p_data)
+{
+    bt_hfp_hf_connection_t* data = (bt_hfp_hf_connection_t*)p_data;
```
Keep name consistent.
Use either `data` or `sal_conn` or `conn`.
Actually I would recommend `sal_hfp_hf_info_t *info` or `zblue_hfp_hf_info_t *info` to distinguish from `struct bt_conn *conn` or `struct bt_hfp_hf *hf`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+static void free_call(void* p_data)
+{
+    bt_hfp_hf_call_t* data = (bt_hfp_hf_call_t*)p_data;
```
`bt_hfp_hf_call_t* call`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    bt_hfp_hf_call_t* data = (bt_hfp_hf_call_t*)p_data;
+    free(data);
+}
+
```
bt_sal_***

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    return !strcmp(sal_call->number, number);
+}
+
+static bt_hfp_hf_call_t *find_call_by_number(bt_hfp_hf_connection_t* conn, const char* number) {
```
Looks like `find_call_by_number` is never used.
Is it possible that `number` is `NULL`?

expliyh replied:
Removed find_call_by_number

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    return (bt_hfp_hf_call_t *)bt_list_find(on_going_calls, sal_bt_hfp_hf_call_number_cmp, (void*)number);
+}
+
+static bt_hfp_hf_call_t *find_call_by_session(bt_hfp_hf_connection_t* conn, struct bt_hfp_hf_call* session) {
```
Can name this a `call` rather than a `session`

`session` is typically a name for Service Level Connection.

gzh-terry replied:
BTW, always use the same name for the same object.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+static bt_hfp_hf_call_t *find_call_in_state(bt_hfp_hf_connection_t* conn, hfp_hf_call_state_t state)
+{
+    if (!conn || !conn->calls) {
```
`find_call_by_state`
And keep the same sanity check for each `find_call_by_***`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+}
+
+static bt_hfp_hf_call_t* build_call(
+    struct bt_hfp_hf_call* session,
```
Check code style.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    }
+
+    if (number) {
+        strncpy(call->number, number, sizeof(call->number) - 1);
```
`strlcpy`

gzh-terry replied:
`strlcpy(call->number, number, sizeof(call->number));`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+    if (number) {
+        strncpy(call->number, number, sizeof(call->number) - 1);
+        call->number[sizeof(call->number) - 1] = '\0';
```
Remove these `'\0'`

expliyh replied:
Func removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+        return;
+    }
+
+    strncpy(call->number, number, sizeof(call->number) - 1);
```
`strlcpy`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    call->type = type;
+}
+
+static bt_hfp_hf_call_t* ensure_call(
```
What is `ensure`?

gzh-terry replied:
It looks like an `new_or_update_call()`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    return NULL;
+}
+
+static bt_hfp_hf_call_t* build_call(
```
`new_call` or `create_call` are more commonly used names.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    uint8_t type;
+    hfp_hf_call_state_t state;
+    struct bt_hfp_hf_call* session;
+} bt_hfp_hf_call_t;
```
`bt_hfp_hf_call` and `bt_hfp_hf_call_t` are quite similar and can be misleading.
Can we use the same structure if they are identical?
Or can we use names that a significantly different?

expliyh replied:
Renamed to bt_hfp_hf_call_info_t

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    return call;
+}
+
+static bt_hfp_hf_connection_t* find_connection_by_call_session(
```
The function name is `find_connection_by_call_session`, then we should input a `call_session` and return a `connection`.
Why there is a `bt_hfp_hf_call_t** call`?

expliyh replied:
In most cases, the call info is used immediately after this function is called, so I output it through the call_info parameter to avoid searching for it twice.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    update_call_indicators(conn);
+}
+
+static bt_hfp_hf_connection_t* find_connection_by_addr(bt_address_t* addr) {
```
Can have a `#define` to reduce pointer jumps.

expliyh replied:
Done with inline

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+static void do_hf_connect(do_hf_connect_params_t *params)
+{
+    struct bt_conn *conn = params->conn;
```
Declare variables first.
```
struct bt_conn *conn = params->conn;
uint16_t channel = params->channel;
struct bt_hfp_hf *hf = NULL;

free(params);
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+static void do_hf_connect(do_hf_connect_params_t *params)
+{
+    struct bt_conn *conn = params->conn;
+    uint16_t channel = params->channel;
```
`uint8_t channel`
Only 5 bits valid.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+        (bt_hfp_hf_connection_t *)zalloc(sizeof(bt_hfp_hf_connection_t));
+    if (!new_connection) {
+        BT_LOGE("%s, Failed to allocate memory for new HFP HF connection", __func__);
+        return;
```
Should we destroy `hf`?

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    }
+
+    bt_hfp_hf_connection_t *new_connection =
+        (bt_hfp_hf_connection_t *)zalloc(sizeof(bt_hfp_hf_connection_t));
```
We may have this `zalloc` before `bt_hfp_hf_connect` to reduce the error return. And may directly set `bt_hfp_hf_connect(conn, &new_connection->hf, channel);`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    new_connection->calls = bt_list_new(free_call);
+    if (!new_connection->calls) {
+        BT_LOGE("%s, Failed to allocate call list for new HFP HF connection", __func__);
+        free(new_connection);
```
May have a `goto error` for all the cases and try to release `hf`

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+    bt_list_add_tail(connections, new_connection);
+
+    BT_LOGD("%s, HFP HF connection established successfully", __func__);
```
It's "connecting" rather than established.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+typedef struct _do_hf_connect_params {
+    struct bt_conn *conn;
+    uint16_t channel;
+} do_hf_connect_params_t;
```
`hf_connect_params_t` is fine.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    return BT_SDP_DISCOVER_UUID_STOP;
+}
+
+static void STACK_CALL(on_hfp_hf_connected)(struct bt_conn *conn, struct bt_hfp_hf *hf)
```
Not really necessary to add `STACK_CALL()` for callbacks.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+uint8_t on_sdp_done(struct bt_conn *conn, struct bt_sdp_client_result *result, const struct bt_sdp_discover_params *ignore)
+{
+    int err;
+    uint16_t value;
```
Name it as `port`?

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+};
+
+typedef struct _bt_hfp_hf_call {
+    char number[32];
```
Use macro for this 32.

expliyh replied:
Removed

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    notify_ring_state(conn, true);
+}
+
+static void STACK_CALL(on_current_call)(struct bt_hfp_hf *hf, struct bt_hfp_hf_current_call *call)
```
Remove `STACK_CALL` for these callbacks.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    }
+
+    uint32_t idx = call->index;
+    hfp_call_direction_t dir = HFP_CALL_DIRECTION_OUTGOING;
```
func `zephyr_dir_to_sal_dir`

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    }
+
+    hfp_hf_call_state_t status = 0;
+    switch (call->status)
```
func `zephyr_call_status_to_sal_call_status`

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+    }
+
+    hfp_call_mpty_type_t mpty = 0;
+    switch (call->multiparty)
```
No need to switch a `bool`.
Just `mpty = call->multiparty ? HFP_CALL_MPTY_TYPE_MULTI : HFP_CALL_MPTY_TYPE_SINGLE`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+{
+(void) hf_features;
+    (void) max_connection;
+    connections = bt_list_new(free_connection);
```
Add more detailed name for global values.
`connections` is hard to be interpretated to a specific type of connection (i.e., ACL or HFP)

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+
+void bt_sal_hfp_hf_cleanup(void)
+{
+    printf("bt_sal_hfp_hf_cleanup: Currently not supported\n");
```
replace `printf` by `BT_LOGD`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,682 @@
+bt_status_t bt_sal_hfp_hf_dial_memory(bt_address_t* addr, uint32_t memory)
+{
+    bt_hfp_hf_connection_t* sal_conn = find_connection_by_addr(addr);
+    char mem_in_str[11];
```
Add macro for 11.

gzh-terry replied:
Actually I would recommend `HFP_PHONENUM_DIGITS_MAX`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/include/sal_hfp_internal.h:
```diff
@@ -0,0 +1,25 @@
+#define CALL_IN_SERVICE(func, data)                   \
+    do {                                              \
+        assert((func) != NULL);                       \
+        do_in_service_loop((service_func_t)(func), (void *)(data)); \
```
Remove `assert` before the final commit.

expliyh replied:
Done

expliyh committed:
zblue: hfp: implement CLCC API and callback

bug: v/78002

Signed-off-by: liyuheng <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+    bt_hfp_hf_connection_t* sal_conn = find_connection_by_addr(addr);
+    if (!sal_conn) {
+        BT_LOGE("%s, Failed to find connection", __func__);
+        return BT_STATUS_FAIL;
```
`BT_STATUS_PARM_INVALID`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+    bt_hfp_hf_call_info_t* incoming = find_call_in_state(sal_conn, HFP_HF_CALL_STATE_INCOMING);
+    if (!incoming) {
+        BT_LOGE("%s, No incoming call to answer", __func__);
+        return BT_STATUS_FAIL;
```
`BT_STATUS_PARM_INVALID` or `BT_STATUS_UNHANDLED`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+        break;
+    }
+
+    hfp_call_mpty_type_t mpty;
```
`mpty == call->multiparty ? HFP_CALL_MPTY_TYPE_MULTI : HFP_CALL_MPTY_TYPE_SINGLE`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+}
+
+static struct bt_hfp_hf_cb hf_callbacks = {
+    .connected = zblue_on_hfp_hf_connected,
```
Keep all the function names consistent, e.g.,
```
.connected = zblue_on_hf_connected,
.incomming = zblue_on_hf_incomming,
.query_call = zblue_on_hf_query_call,
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+    bt_sal_get_remote_address(sal_conn->conn, bd_addr);
+
+    if (!call) {
+        BT_LOGD("%s, CLCC Finished", __func__);
```
`BT_ADDR_LOG("CLCC finished from %s", bd_addr);`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+        return;
+    }
+
+    bt_address_t* bd_addr = zalloc(sizeof(bt_address_t));
```
Do not use `zalloc`, just
```
bt_address_t bd_addr;
bt_sal_get_remote_address(sal_conn->conn, &bd_addr);
```

Then we can:
1. reduce memory fragments.
2. no need to free memory manually.
3. less cost on memory and cycles, especially in debug mode.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/include/sal_hfp_internal.h:
```diff
@@ -0,0 +1,25 @@
+
+#define CALL_IN_SERVICE(func, data)                   \
+    do {                                              \
+        assert((func) != NULL);                       \
```
Remove `assert` before final commit.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_ag_interface.c:
```diff
@@ -0,0 +1,221 @@
+
+bt_status_t bt_sal_hfp_ag_connect(bt_address_t* addr)
+{
+
```
Code style.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+
+static bt_list_t* connections = NULL;
+
+NET_BUF_POOL_DEFINE(sdp_discover_pool, CONFIG_BT_MAX_CONN, BT_L2CAP_BUF_SIZE(CONFIG_BT_L2CAP_TX_MTU),
```
Remember adding this pool to `zblue/port/sections/defines.c`

gzh-terry replied:
BTW, use a name that specified for HF, otherwise, use the pool defined by someone else.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+
+static __attribute__((unused)) void free_call(void* p_data)
+{
+    bt_hfp_hf_call_info_t* data = (bt_hfp_hf_call_info_t*)p_data;
```
```
bt_hfp_hf_call_info_t* sal_call= (bt_hfp_hf_call_info_t*)p_data;
free(sal_call);
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+        BT_LOGE("%s, calls is NULL", __func__);
+    }
+
+    bt_list_t* on_going_calls = conn->calls;
```
`on_going` -> `ongoing`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+    return call;
+}
+
+static bt_hfp_hf_call_info_t* ensure_call(
```
`ensure` is not very frequently used, try some other names by deepseek.
My suggestion:
```
ensure_call -> find_or_create_call
build_call -> new_call
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,637 @@
+    hfp_hf_on_connection_state_changed(&bd_addr, PROFILE_STATE_CONNECTED, 0, 0);
+}
+
+static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call* call)
```
Use either `call` or `z_context`, not mixed.

gzh-terry replied:
I would recomment to keep `call` and `sal_call` only.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,636 @@
+            params->channel = value;
+            CALL_IN_SERVICE(do_hf_connect, params);
+            params = NULL;
+            if (err != 0) {
```
This would never happens since we have `err = 0` at line 294.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,636 @@
+    if (bt_sal_get_remote_address(conn, &bd_addr) != BT_STATUS_SUCCESS)
+        return;
+    if (!find_connection_by_addr(&bd_addr)) {
+        bt_hfp_hf_connection_t* new_connection = (bt_hfp_hf_connection_t*)zalloc(sizeof(bt_hfp_hf_connection_t));
```
Move these common codes into:
```
static bt_hfp_hf_connection_t* new_hf_connection(struct bt_conn* conn, struct bt_hfp_hf* hf)
{
    bt_hfp_hf_connection_t* sal_conn = (bt_hfp_hf_connection_t*)zalloc(sizeof(bt_hfp_hf_connection_t));
    bt_sal_get_remote_address(conn, &sal_conn->addr);
    sal_conn->conn = conn;
    sal_conn->hf = hf;

    /** other initializations */

}
```

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,636 @@
+
+#define STACK_CALL(func) zblue_##func
+
+static bt_list_t* connections = NULL;
```
`g_sal_hf_conn = NULL;`

expliyh replied:
Renamed to g_sal_hf_conn_list

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,636 @@
+            0);
+        return;
+    } else {
+        BT_LOGD("%s, CLCC %d: %s", __func__, call->index, call->number);
```
Same here, print address.

expliyh replied:
Not do for addr printed when finish. if print here, log will too long

expliyh committed:
zblue: hfp: implement CNUM API and callback.

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement disconnect API and callback

bug: v/77997

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement reject call api

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement call rejected callback.

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement call accept callback.

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement outgoing call callback

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement call terminate callback.

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement call held callback.

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement call retrieve callback

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement CLIP callback

bug: v/77999

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement vgm and vgs callback

bug: v/77999

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement codec_negotiate callback

bug: v/77999

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement voice_recognition callback

bug: v/77999

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement remote_ringing callback

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement ring_indication callback

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_connect_audio API

bug: v/80257

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement audio connect/disconnect API and callback

bug: v/80257

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+    return !bt_addr_compare(&sal_conn->addr, addr);
+}
+
+static bool bt_hfp_hf_call_context_cmp(void* sal_context, void* z_context)
```
sal_call_context_cmp

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+    return sal_call->context == call;
+}
+
+static bt_hfp_hf_call_info_t* find_call_by_context(bt_hfp_hf_connection_t* conn, struct bt_hfp_hf_call* z_context)
```
Input arg `conn` -> `sal_conn`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+    return (bt_hfp_hf_call_info_t*)bt_list_find(call_list, bt_hfp_hf_call_context_cmp, z_context);
+}
+
+static bt_hfp_hf_call_info_t* find_call_by_state(bt_hfp_hf_connection_t* conn, hfp_hf_call_state_t state)
```
`conn` -> `sal_conn`

expliyh replied:
Done

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_hold_call API

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+        return NULL;
+    }
+
+    call->type = 0;
```
Either have a macro for this `0`, or remove it since we already have `zalloc`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+}
+
+static bt_hfp_hf_call_info_t* find_or_create_call(
+    bt_hfp_hf_connection_t* conn,
```
Line feed at 100 characters.

gzh-terry replied:
Try this

<img width="1050" height="484" alt="Image" src="https://github.com/user-attachments/assets/90dbeca2-060b-48c1-bc18-8fa5d7e2f54c" />

expliyh replied:
Done

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_hangup_call API

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_call_control API

bug: v/80256

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_set_volume API

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_start/stop_voice_recognition API

bug: v/80257

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_send_battery_level API

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,613 @@
+        return NULL;
+    }
+
+    bt_hfp_hf_call_info_t* call = find_call_by_context(conn, z_context);
```
`conn` -> `sal_conn`

expliyh replied:
Done

expliyh committed:
zblue: hfp: implement bt_sal_hfp_hf_send_dtmf API

bug: v/78000

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: disconnect profile when sdp fails

bug: v/77997
Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: implement vendor_specific API and callback

bug: v/78306

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: add command complete queue

bug: v/78600

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: fix: crash when connect

bug: v/80258

Rootcause: unref connection context before use

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: redial when bt_sal_hfp_hf_dial_number was called with a NULL number

bug: v/80258

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: fix crash when connect after disconnect.

bug: v/80258

Rootcause: unnesessary unref.

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_ag_interface.c:
```diff
@@ -0,0 +1,223 @@
+
+bt_status_t bt_sal_hfp_ag_connect(bt_address_t* addr)
+{
+
```
Remove line break.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+        sal_conn->pending_completes = NULL;
+    }
+    sal_conn->conn = NULL;
+    sal_conn->sco_conn = NULL;
```
Redundant before `free(sal_conn)`

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+            return BT_SDP_DISCOVER_UUID_STOP;
+        }
+        BT_LOGE("%s, remote device does not support HFP AG feature", __func__);
+        bt_conn_unref(conn);
```
Where do we get this `ref`?
If this `unref` is reasonable, remember to unref at `do_hf_connect` as well, in case of anything failed.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+    }
+    conn->sco_conn = sco_conn;
+    if (conn->sco_conn) {
+        bt_conn_ref(conn->sco_conn);
```
Remember to `bt_conn_unref` on disconnected if we `bt_conn_ref` here

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+    enum bt_at_result result,
+    enum bt_at_cme err)
+{
+
```
Remove line break.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+        return BT_STATUS_NOMEM;
+    }
+
+    char cmd_buf[HFP_AT_LEN_MAX + 1];
```
Use heap memory, this is too large for stack.

expliyh replied:
Done

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1346 @@
+    char cmd_buf[HFP_AT_LEN_MAX + 1];
+    size_t copy_len = (len < HFP_AT_LEN_MAX) ? len : HFP_AT_LEN_MAX;
+    memcpy(cmd_buf, cmd, copy_len);
+    cmd_buf[copy_len] = '\0';
```
BTW, why do we do this copy? can we use `cmd` it self.
`cmd` should have a terminating `\0` since it's a `char*`

expliyh replied:
Done

expliyh committed:
zblue: hfp: hf: add command complete callback

bug: v/78600

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: implement cleanup api

bug: v/77997

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: fix ARRAY_SIZE redefine warning.

bug: v/80258

Rootcause: ARRAY_SIZE redefined

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

chengkai15 commented on service/stacks/zephyr/include/sal_hfp_internal.h:
```diff
@@ -0,0 +1,24 @@
+
+#include "service_loop.h"
+
+#define Z_API(func) z_##func
```
remove

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/include/sal_hfp_internal.h:
```diff
@@ -0,0 +1,24 @@
+
+#define Z_API(func) z_##func
+
+#define CALL_IN_SERVICE(func, data)                                \
```
move to common head for all profile

expliyh replied:
File removed

chengkai15 commented on service/stacks/zephyr/sal_hfp_ag_interface.c:
```diff
@@ -13,3 +13,211 @@
+
+bt_status_t bt_sal_hfp_ag_disconnect(bt_address_t* addr)
+{
+    return BT_STATUS_UNSUPPORTED;
```
add addr unsed

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -38,7 +38,140 @@
+    hfp_callheld_t held_state;
+} bt_hfp_hf_connection_t;
+
+static __attribute__((unused)) void free_connection(void* data)
```
why do we add __attribute__((unused))?

expliyh replied:
removed

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -38,7 +38,140 @@
+        return NULL;
+    }
+
+    bt_list_node_t* node;
```
move var declear to the header of func （LINE:111）

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -38,7 +38,140 @@
+        }
+
+        bt_hfp_hf_call_info_t* sal_call = find_call_by_context(conn, z_context);
+        if (sal_call) {
```
if（sal_call && call_info）

expliyh replied:
Not changed. There are codes to run when sal_call != NULL and call_info == NULL

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+    uint8_t channel = params->channel;
+    struct bt_hfp_hf* hf = NULL;
+
+    free(params);
```
could we free params at caller？

expliyh replied:
Not changed: these codes will be removed when adapting direct profile connect/disconnect.
And this fuction is designed to be called in service loop,  params can't be freed at caller.

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+        return;
+    }
+
+    new_hf_connection(conn, hf);
```
could we check new_hf_connection fail？

expliyh replied:
Done in next patch

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+    BT_LOGD("%s, HFP HF connecting", __func__);
+}
+
+uint8_t zblue_on_sdp_done(struct bt_conn* conn, struct bt_sdp_client_result* result, const struct bt_sdp_discover_params* ignore)
```
add static for Local func

expliyh replied:
Done in next patch

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+    int err;
+    uint16_t port;
+
+    if (result->resp_buf != NULL) {
```
could we simple it with many if and return.  like:
if(...) {
...
return XX
}


if(...) {
...
return XX
}

expliyh replied:
Done in next patch

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+static void zblue_on_connected(struct bt_conn* conn, struct bt_hfp_hf* hf)
+{
+    bt_address_t bd_addr;
+    if (bt_sal_get_remote_address(conn, &bd_addr) != BT_STATUS_SUCCESS) {
```
we'd better add line break between func var declare and body

expliyh replied:
Done in next patch

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -171,8 +182,81 @@ static __attribute__((unused)) bt_hfp_hf_connection_t* new_hf_connection(struct
+    if (bt_sal_get_remote_address(conn, &bd_addr) != BT_STATUS_SUCCESS) {
+        return;
+    }
+    if (!find_connection_by_addr(&bd_addr)) {
```
we'd better add line break after parentheses

expliyh replied:
Done in next patch

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -207,8 +291,12 @@ static __attribute__((unused)) struct bt_hfp_hf_cb hf_callbacks = {
+    (void)hf_features;
+    (void)max_connection;
+    g_sal_hf_conn_list = bt_list_new(free_connection);
+    SAL_CHECK_RET(Z_API(bt_hfp_hf_register)(&hf_callbacks), 0);
```
if register fail,  g_sal_hf_conn_list would be a wild pointer

expliyh replied:
Fixed

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -217,14 +305,24 @@ void bt_sal_hfp_hf_cleanup(void)
+
+    BT_LOGD("%s, Start SDP discovery for HFP HF", __func__);
+
+    int ret = bt_sdp_discover(conn, &sdp_discover);
```
please add var declare to func head

expliyh replied:
Code removed

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -182,6 +217,22 @@ static bt_hfp_hf_connection_t* new_hf_connection(struct bt_conn* conn, struct bt
+
+    sal_call->state = state;
+
+    if (state == HFP_HF_CALL_STATE_DISCONNECTED && sal_conn && sal_conn->calls) {
```
could we decoupling calls remove from set_call_state?  which was a bad design

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -115,6 +115,24 @@ static bt_hfp_hf_call_info_t* find_call_by_context(bt_hfp_hf_connection_t* sal_c
+        return NULL;
+    }
+
+    bt_list_node_t* node;
```
dito

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -343,6 +343,86 @@ static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call*
+
+    if (!call) {
+        BT_ADDR_LOG("CLCC finished from %s", &bd_addr);
+        hfp_hf_on_current_call_response(
```
please format the codes

expliyh replied:
Already format using clang-format

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -343,6 +343,86 @@ static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call*
+        dir = HFP_CALL_DIRECTION_INCOMING;
+        break;
+    default:
+        BT_LOGW("%s, Unknown direction: %d", __func__, call->dir);
```
could we return for exception case？

expliyh replied:
I‘d like to not return to keep at least call list sync with stack

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -343,6 +343,86 @@ static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call*
+    hfp_call_mpty_type_t mpty = call->multiparty ? HFP_CALL_MPTY_TYPE_MULTI : HFP_CALL_MPTY_TYPE_SINGLE;
+
+    hfp_hf_on_current_call_response(
+        &bd_addr,
```
dito

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -343,6 +343,86 @@ static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call*
+        break;
+    }
+
+    hfp_hf_call_state_t status = 0;
```
could we replace with unkown status？

expliyh replied:
No unknown status in enum

expliyh committed:
zblue: hfp: hf: fix SCO context not found when disconnect

bug: v/80025

Rootcause: SCO context not recorded when connect

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/include/sal_hfp_ag_interface.h:
```diff
@@ -0,0 +1,55 @@
+bt_status_t bt_sal_hfp_ag_send_at_cmd(bt_address_t* addr, const char* atcmd, uint16_t length);
+bt_status_t bt_sal_hfp_ag_manufacture_id_response(bt_address_t* addr,
+    const char* manufacturer_id,
+    uint16_t length);
```
Line feed on 100 characters.

expliyh committed:
zblue: hfp: hf: fix hangup fail when multicall using

bug: v/80184

Rootcause: Use terminate stact API which is designed for single call control when multicall using

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: fix crash when enable bluetooth after disable

bug: v/80021

Rootcause: Not unregister callbacks when cleanup.
Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

expliyh committed:
zblue: hfp: hf: profile direct connect

bug: v/77997

Signed-off-by: YuhengLi <liyuheng@xiaomi.com>

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1335 @@
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/classic/hfp_hf.h>
+#include <zephyr/bluetooth/classic/sdp.h>
+#include <zephyr/bluetooth/conn.h>
```
Remove unused headers, e.g., `conn`, `l2cap`, `net_buf`, `atomic`

expliyh replied:
Done

gzh-terry reviewed (approved):
Generally fine to me, just a few comments.

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1335 @@
+        return BT_STATUS_FAIL;
+    }
+
+    SAL_CHECK_RET(Z_API(bt_hfp_hf_disconnect)(sal_conn->hf), 0);
```
call `bt_sal_profile_disconnect_request` here.

expliyh replied:
Done

chengkai15 commented on Makefile:
```diff
@@ -258,6 +258,12 @@ endif #CONFIG_BLUETOOTH_GATT_CLIENT
 ifeq ($(CONFIG_BLUETOOTH_GATT_SERVER), y)
 	CSRCS += service/stacks/zephyr/sal_gatt_server_interface.c
 endif #CONFIG_BLUETOOTH_GATT_SERVER
+ifeq ($(CONFIG_BLUETOOTH_HFP_HF), y)
```
move to classic subclass?

expliyh replied:
Done

chengkai15 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -367,6 +367,33 @@ static void zblue_on_incoming_call(struct bt_hfp_hf* hf, struct bt_hfp_hf_call*
+
+    hfp_subscriber_number_service_t fw_service = 0;
+    switch (service) {
+    case 4:
```
could we replace the numbers with  macro？

expliyh replied:
<img width="932" height="787" alt="image" src="https://github.com/user-attachments/assets/2189e2ac-9caf-42ee-8f4a-6e3cb6572c03" />
Zephyr uses uint8_t, no marco defined

gzh-terry commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+    }
+
+    /** Step 1: SDP discovery */
+    if (g_conn_params == NULL) {
```
TODO: remove `g_conn_params` in the future once `do_hf_connect` can carry a `*userdata`

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+        return NULL;
+    }
+
+    call->index = 0;
```
zalloc has init these to 0

expliyh replied:
Done

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+    return NULL;
+}
+
+static inline __attribute__((unused)) bt_hfp_hf_connection_t* find_connection_by_conn(struct bt_conn* conn)
```
why use __attribute__((unused)), just remove this func

expliyh replied:
Done

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+    }
+
+    /** Step 2: SLC initiating */
+    uint8_t channel = g_conn_params->channel;
```
move these to the header of func, and check all of these.

expliyh replied:
Done

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+        return BT_STATUS_FAIL;
+    }
+
+    SAL_CHECK_RET(Z_API(bt_hfp_hf_disconnect)(sal_conn->hf), 0);
```
report DISCONNECTING?

expliyh replied:
Listed as TODO

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+    }
+    bt_address_t* bd_addr = &conn->addr;
+
+    hfp_hf_on_connection_state_changed(bd_addr, PROFILE_STATE_DISCONNECTING, 0, 0);
```
why report disconnecting here, not in disconenct req

expliyh replied:
Listed as TODO

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+        sal_call->type = type;
+    }
+
+    const char* num = number ? number : "";
```
the number comes from zblue, it's better to memcpy to upper

expliyh replied:
Done

zhangyuan376 commented on service/stacks/zephyr/sal_hfp_hf_interface.c:
```diff
@@ -0,0 +1,1351 @@
+    if (!call) {
+        BT_ADDR_LOG("CLCC finished from %s", &bd_addr);
+        hfp_hf_on_current_call_response(
+            &bd_addr,
```
these code can merge to line 829

expliyh replied:
Done
